File: python/src/part1/chapter3/tree.py
"""Tree."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Generic, TypeVar, TypeAlias, Callable

A = TypeVar("A")
B = TypeVar("B")


class Tree(Generic[A]):
    """Tree"""

    @property
    def pattern(self) -> SubType[A]:
        raise NotImplementedError()

    def size(self) -> int:
        match self.pattern:
            case Leaf():
                return 1
            case Branch(left, right):
                return 1 + left.size() + right.size()

    def maximum(self: Tree[int]) -> int:
        match self.pattern:
            case Leaf(value):
                return value
            case Branch(left, right):
                return max(left.maximum(), right.maximum())

    def depth(self) -> int:
        match self.pattern:
            case Leaf():
                return 0
            case Branch(left, right):
                return 1 + max(left.depth(), right.depth())

    def map(self, f: Callable[[A], B]) -> Tree[B]:
        match self.pattern:
            case Leaf(value):
                return Leaf(f(value))
            case Branch(left, right):
                return Branch(left.map(f), right.map(f))


@dataclass(frozen=True)
class Leaf(Tree[A]):
    """Leaf"""

    value: A

    @property
    def pattern(self) -> SubType[A]:
        return self


@dataclass(frozen=True)
class Branch(Tree[A]):
    """Branch"""

    left: Tree[A]
    right: Tree[A]

    @property
    def pattern(self) -> SubType[A]:
        return self


SubType: TypeAlias = Leaf[A] | Branch[A]

File: typescript/src/part1/chapter3/tree.ts
import { match, P } from "ts-pattern";

export type Tree<T> = Leaf<T> | Branch<T>
export interface Leaf<T> {
    type: 'Leaf'
    value: T
}
export interface Branch<T> {
    type: 'Branch'
    left: Tree<T>
    right: Tree<T>
}

const max = (left: number, right: number): number => left > right ? left : right

const size = <T>(t: Tree<T>): number =>
    match(t)
        .with({ type: 'Leaf' }, () => 1)
        .with({ type: 'Branch' }, (branch) => 1 + size(branch.left) + size(branch.right))
        .exhaustive()

export const maximum = (t: Tree<number>): number =>
    match(t)
        .with({ type: 'Leaf' }, (leaf) => leaf.value)
        .with({ type: 'Branch' }, (branch) => {
            const left = maximum(branch.left)
            const right = maximum(branch.right)
            return left > right ? left : right
        })
        .exhaustive()

export const depth = <T>(t: Tree<T>): number =>
    match(t)
        .with({ type: 'Leaf' }, () => 0)
        .with({ type: 'Branch' }, (branch) => 1 + max(depth(branch.left), depth(branch.right)))
        .exhaustive()

export const map = <T, B>(t: Tree<T>, f: (t: T) => B): Tree<B> =>
    match(t)
        .with({ type: 'Leaf' }, (leaf) => ({ ...leaf, value: f(leaf.value) } as Tree<B>))
        .with({ type: 'Branch' }, (branch) => ({ ...branch, left: map(branch.left, f), right: map(branch.right, f) } as Tree<B>))
        .exhaustive()

const fold = <A, B>(t: Tree<A>, f: (a: A) => B, g: (l: B, r: B) => B): B =>
    match(t)
        .with({ type: 'Leaf' }, (leaf) => f(leaf.value))
        .with({ type: 'Branch' }, (branch) => g(fold(branch.left, f, g), fold(branch.right, f, g)))
        .exhaustive()

export const sizeViaFold = <A>(t: Tree<A>): number =>
    fold(t, (_) => 1, (l, r) => 1 + l + r)

export const depthViaFold = <A>(t: Tree<A>): number =>
    fold(t, (_) => 0, (left, right) => 1 + max(left, right))

export const mapViaFold = <A, B>(t: Tree<A>, f: (a: A) => B): Tree<B> =>
    fold(t, (value) => ({ type: 'Leaf', value: f(value) } as Tree<B>), (left, right) => ({ type: 'Branch', left: left, right: right } as Tree<B>))
